src/main/java/com/rapid7/client/dcerpc/msrrp/RegistryService.java

```diff
@@ -171,21 +200,54 @@ protected String canonicalize(String keyPath) {
         }
     }
 
-    protected byte[] openKey(final String hiveName, final String keyPath) throws IOException {
+    public byte[] openKey(final String hiveName, final String keyPath) throws IOException {
+        return openKey(hiveName, keyPath, ACCESS_MASK);
+    }
+
+    public byte[] openKey(final String hiveName, final String keyPath, EnumSet<AccessMask> access) throws IOException {
         final String canonicalizedKeyPath = canonicalize(keyPath);
         if (canonicalizedKeyPath.isEmpty()) {
             return openHive(hiveName);
         }
         synchronized (keyPathCache) {
-            if (keyPathCache.containsKey(canonicalizedKeyPath)) {
-                return keyPathCache.get(canonicalizedKeyPath);
+            final RegistryHandleKey cachingKey = new RegistryHandleKey(canonicalizedKeyPath, access);
+            if (keyPathCache.containsKey(cachingKey)) {
+                return keyPathCache.get(cachingKey);
             }
             final byte[] hiveHandle = openHive(hiveName);
-            final BaseRegOpenKey request = new BaseRegOpenKey(hiveHandle, canonicalizedKeyPath, 0, ACCESS_MASK);
+            final BaseRegOpenKey request = new BaseRegOpenKey(hiveHandle, canonicalizedKeyPath, 0, access);
             final HandleResponse response = callExpectSuccess(request, "BaseRegOpenKey");
             final byte[] keyHandle = response.getHandle();
-            keyPathCache.put(canonicalizedKeyPath, keyHandle);
+            keyPathCache.put(cachingKey, keyHandle);
             return keyHandle;
         }
     }
+
+    private static class RegistryHandleKey {
+        private final String path;
+        private final int access;
+
+        RegistryHandleKey(final String path, final EnumSet<AccessMask> access) {
+            this(path, (int) EnumUtils.toLong(access));
+        }
+
+        RegistryHandleKey(final String path, final long access) {
+            this.path = path;
+            this.access = (int) access;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (!(o instanceof RegistryHandleKey))
+                return false;
+            final RegistryHandleKey other = (RegistryHandleKey) o;
+            return path.equals(other.path) && access == other.access;
```

Comment:
This won't handle nulls gracefully, consider `java.util.Objects`:
return Objects.equals(path, other.path) && access == other.access;

Reply:
I added a null check in the constructor. This cannot be null. Canonicalization changes null the empty string.